test/test_jit_cuda_fuser.py

```diff
@@ -86,20 +86,57 @@ def t(x, y, z, q):
     @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING, "Requires profiling node to run cuda fuser")
     @skipIfRocm
     def test_scalar_input(self):
-        def t(x, y, z):
-            # type: (Tensor, Tensor, float) -> Tensor
+        def t(x : torch.Tensor, y : torch.Tensor, z : float):
             o = x + y
             o = o + z
             return o
         t_jit = torch.jit.script(t)
-        x = torch.randn(4, 8, dtype=torch.float, device="cuda")
-        y = torch.randn(4, 8, dtype=torch.float, device="cuda")
+        x = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda")
+        y = torch.randn(4, 8, 1, 32, dtype=torch.float, device="cuda")
+        y = y.expand(4, 8, 32, 32)
+        jit_o = t_jit(x, y, 2.0)
+        jit_o = t_jit(x, y, 2.0)
+        o = t(x, y, 2.0)
+        self.assertEqual(o, jit_o)
+        self.assertTrue(self._has_cuda_fusion_group(t_jit.graph_for(x, y, 2.0)))
+
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING, "Requires profiling node to run cuda fuser")
+    @skipIfRocm
+    def test_broadcasting(self):
+        def t(x : torch.Tensor, y : torch.Tensor, z : float):
+            o = x + y
+            o = o + z
+            return o
+        t_jit = torch.jit.script(t)
+        x = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda")
+        y = torch.randn(32, 32, dtype=torch.float, device="cuda")
         jit_o = t_jit(x, y, 2.0)
         jit_o = t_jit(x, y, 2.0)
         o = t(x, y, 2.0)
         self.assertEqual(o, jit_o)
         self.assertTrue(self._has_cuda_fusion_group(t_jit.graph_for(x, y, 2.0)))
 
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING, "Requires profiling node to run cuda fuser")
+    @skipIfRocm
+    def test_broadcasting_multiple_output_shape(self):
```

Comment:
What does the JIT IR look like in this case?

Reply:
it's a graph with branches and joined at the end.

```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %z.1 : Tensor):
  %16 : None = prim::Constant()
  %5 : int = prim::Constant[value=1]()
  %4 : int = prim::Constant[value=12]() # test_jit_cuda_fuser.py:125:20
  %o.1 : Tensor = aten::add(%x.1, %4, %5) # test_jit_cuda_fuser.py:125:16
  %o1.1 : Tensor = aten::add(%o.1, %y.1, %5) # test_jit_cuda_fuser.py:126:17
  %o2.1 : Tensor = aten::add(%o.1, %z.1, %5) # test_jit_cuda_fuser.py:127:17
  %17 : Tensor = aten::sum(%o1.1, %16) # test_jit_cuda_fuser.py:128:17
  %20 : Tensor = aten::sum(%o2.1, %16) # test_jit_cuda_fuser.py:128:28
  %oo.1 : Tensor = aten::add(%17, %20, %5) # test_jit_cuda_fuser.py:128:17
  return (%oo.1)
```

After profiling, if you ignore the bailout node, you can say all three `add` node has different output size, hence they are not fused with each other.
```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %z.1 : Tensor):
  %5 : None = prim::Constant()
  %4 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=12]() # test_jit_cuda_fuser.py:125:20
  %37 : int = prim::BailoutTemplate_0()
  %32 : Float(4, 32, 32) = prim::BailOut[index=0](%37, %z.1, %x.1, %y.1)
  %33 : Float(2, 32, 32) = prim::BailOut[index=1](%37, %y.1, %x.1, %32)
  %34 : Float(32, 32) = prim::BailOut[index=2](%37, %x.1, %33, %32)
  %o.1 : Float(32, 32) = aten::add(%34, %3, %4) # test_jit_cuda_fuser.py:125:16
  %o1.1 : Float(2, 32, 32) = aten::add(%o.1, %33, %4) # test_jit_cuda_fuser.py:126:17
  %o2.1 : Float(4, 32, 32) = aten::add(%o.1, %32, %4) # test_jit_cuda_fuser.py:127:17
  %15 : Tensor = aten::sum(%o1.1, %5) # test_jit_cuda_fuser.py:128:17
  %35 : Float() = prim::BailOut[index=3](%37, %15, %o2.1)
  %17 : Tensor = aten::sum(%o2.1, %5) # test_jit_cuda_fuser.py:128:28
  %36 : Float() = prim::BailOut[index=4](%37, %17, %35)
  %oo.1 : Float() = aten::add(%35, %36, %4) # test_jit_cuda_fuser.py:128:17
  return (%oo.1)
```